### PR TITLE
Python: Add support for threat models

### PR DESCRIPTION
**Pull request by RasmusWL**

### CODE REVIEW

File: python/ql/test/library-tests/frameworks/stdlib/threat_models.py [hunk 1]
```diff
@@ -30,14 +30,14 @@
...
+args = parser.parse_args() # $ threatModelSource[commandargs]=parser.parse_args()
+ensure_tainted(args.foo) # $ tainted
 
 explicit_argv_parsing = parser.parse_args(sys.argv) # $ threatModelSource[commandargs]=sys.argv
-ensure_tainted(explicit_argv_parsing.foo) # $ MISSING: tainted
+ensure_tainted(explicit_argv_parsing.foo) # $ tainted
 
 fake_args = parser.parse_args(["<foo>"])
-ensure_not_tainted(fake_args.foo)
+ensure_not_tainted(fake_args.foo) # $ SPURIOUS: tainted
```
Comment (asgerf) [inline python/ql/test/library-tests/frameworks/stdlib/threat_models.py:40]: Any idea why this got tainted?

**Review by asgerf — COMMENTED** [review]
Looks great! I mainly a comment about the class names otherwise the structure LGTM

File: python/ql/lib/semmle/python/Concepts.qll [hunk 2]
```diff
@@ -10,6 +10,54 @@ private import semmle.python.dataflow.new.RemoteFlowSources
...
 private import semmle.python.security.internal.EncryptionKeySizes
+private import codeql.threatmodels.ThreatModels
+
+/**
+ * A data flow source, for a specific threat-model.
+ *
+ * Extend this class to refine existing API models. If you want to model new APIs,
+ * extend `ThreatModelSource::Range` instead.
+ */
+class ThreatModelSource extends DataFlow::Node instanceof ThreatModelSource::Range {
```
[the remark this answers is not in the document]
Comment (asgerf) [reply python/ql/lib/semmle/python/Concepts.qll:21]: Whatever we do we ought to treat the naming of the two classes as a single decision; not something where we try to make a decision for each class in isolation.

The proposed rename to `FlowSource` made sense when the other class would be called `ThreatModelFlowSource`; but it doesn't work so so nicely with `ActiveThreatModelSource`. I'd prefer the combination `ThreatModelSource/ActiveThreatModelSource` over `FlowSource/ActiveThreatModelSource`.

[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/lib/semmle/python/Concepts.qll:21]: Alright, I'm feeling strongly in favor of `ActiveThreatModelSource`, so it seems like I won't add the `FlowSource` commit :+1:

File: python/ql/lib/semmle/python/Concepts.qll [hunk 3]
```diff
@@ -10,6 +10,62 @@ private import semmle.python.dataflow.new.RemoteFlowSources
...
+}
+
+/**
+ * A data flow source that is enabled in the current threat model configuration.
+ */
+class ActiveThreatModelSource extends DataFlow::Node {
+  ActiveThreatModelSource() {
+    exists(string kind |
+      currentThreatModel(kind) and
+      this.(ThreatModelSource).getThreatModel() = kind
```
Comment (tausbn) [inline python/ql/lib/semmle/python/Concepts.qll:65]: Silly question, but why doesn't this extend `ThreatModelSource`? After all, it only holds for instances of that class anyway.

**Review by tausbn — DISMISSED** [review]
Just one minor comment, otherwise I think this looks really good! (I'm approving it now in case the minor comment is irrelevant.)

File: python/ql/lib/semmle/python/Concepts.qll (hunk 3 above)
Comment (RasmusWL) [reply python/ql/lib/semmle/python/Concepts.qll:65]: I copied this code from the static languages, and didn't adjust things. Will do a fixup :+1:

**Review by tausbn — APPROVED** [review]

**Review by tausbn — APPROVED** [review]
:shipit:

**Pull request merged**